editor/src/messages/tool/common_functionality/shape_editor.rs

```diff
@@ -201,6 +246,106 @@ impl ShapeState {
+		let subpath = &vector_data.subpaths[subpath_index];
+		let manipulator_groups = subpath.manipulator_groups();
+
+		match group_data.selection_type {
+			OUT_HANDLE | ANCHOR_OUT => {
+				blink(ManipulatorPointId::new(group_data.group_id, SelectedType::OutHandle));
+			}
+			IN_HANDLE | ANCHOR_IN => {
+				blink(ManipulatorPointId::new(group_data.group_id, SelectedType::InHandle));
+			}
+			ANCHOR | BOTH_HANDLES | ALL => {
+				let anchor_position = SelectedType::Anchor.get_position(group)?;
+				let out_position = SelectedType::OutHandle.get_position(group)?;
+				let in_position = SelectedType::InHandle.get_position(group)?;
+
+				let anchor_out = anchor_position - out_position;
+				let anchor_out_norm = anchor_out.normalize_or_zero();
+				let anchor_in = anchor_position - in_position;
+				let anchor_in_norm = anchor_in.normalize_or_zero();
+
+				// Force the bisectors into an isosceles triangle so they are easier
+				// to get a perpendicular vector out of.
+				let perpendicular_prenorm = anchor_out_norm - anchor_in_norm;
+				// We need to align the perpendicular vector with the flow of the
+				// adjacent subpaths when the handles are identical, we use this variable
+				// to do that.
+				let mut flip_in_and_out = 1.0;
+
+				let perpendicular = if anchor_out.abs_diff_eq(anchor_in, 1e-10) {
+					// Flip the vector if it is not facing towards the same direction as the anchor
+					let handle_vector_perp = anchor_out_norm.perp();
+
+					let previous_position = index.checked_sub(1).and_then(|index| manipulator_groups.get(index)).map(|group| group.anchor);
+					let next_position = manipulator_groups.get(index + 1).map(|group| group.anchor);
+
+					if previous_position.filter(|&pos| (pos - anchor_position).normalize().dot(handle_vector_perp) < 0.).is_some()
+						|| next_position.filter(|&pos| (pos - anchor_position).normalize().dot(handle_vector_perp) > 0.).is_some()
+					{
+						flip_in_and_out = -1.0;
+					}
+
+					handle_vector_perp
+				} else {
+					(perpendicular_prenorm).normalize_or_zero()
+				};
+
+				blink(ManipulatorPointId::new(group_data.group_id, SelectedType::Anchor));
+
+				responses.add(GraphOperationMessage::Vector {
+					layer: group_data.layer_id.clone(),
+					modification: VectorDataModification::SetManipulatorPosition {
+						point: ManipulatorPointId::new(group_data.group_id, SelectedType::OutHandle),
+						position: anchor_position - (flip_in_and_out * perpendicular * anchor_out.length()),
+					},
+				});
+
+				responses.add(GraphOperationMessage::Vector {
+					layer: group_data.layer_id.clone(),
+					modification: VectorDataModification::SetManipulatorPosition {
+						point: ManipulatorPointId::new(group_data.group_id, SelectedType::InHandle),
+						position: anchor_position + (flip_in_and_out * perpendicular * anchor_in.length()),
+					},
+				});
+			}
+			_ => {
+				warn!("The current selection type is invalid");
+			}
+		};
+
+		Some(())
+	}
```

Comment:
We already have a function for changing to and from sharp called `flip_sharp`. Perhaps this could be extended to support set sharp to remove code duplication?

Reply:
I might be able to factor `flip_sharps` smoothing code into a separate function to reduce duplication. But I think the functions don't have enough control flow in common for unifying them into one function to make sense.

Reply:
They both have the code for removing or creating the handles and detecting if a manipulator group is already sharp. These should probably be separate functions which can be called for both behaviours.

Reply:
all the Shared code between the two functions has been moved into a separate, shared, function.